File: server/src/main/java/de/ingrid/igeserver/api/BehavioursApiController.java
package de.ingrid.igeserver.api;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.orientechnologies.orient.core.db.ODatabaseSession;
import de.ingrid.igeserver.db.DBApi;
import de.ingrid.igeserver.services.MapperService;
import de.ingrid.igeserver.utils.AuthUtils;
import de.ingrid.igeserver.utils.DBUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.security.Principal;
import java.util.List;
import java.util.stream.Collectors;

@RestController
@RequestMapping(path = "/api")
public class BehavioursApiController implements BehavioursApi {

    private static Logger log = LogManager.getLogger(BehavioursApiController.class);

    @Autowired
    private DBApi dbService;

    @Autowired
    private DBUtils dbUtils;

    @Autowired
    private AuthUtils authUtils;

    public ResponseEntity<List<JsonNode>> getBehaviours(Principal principal) throws ApiException {
        System.out.println(principal == null ? "principal is null" : "principal is " + principal.getName());

        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {

            List<JsonNode> behaviours = this.dbService.findAll(DBApi.DBClass.Behaviours.name());

            // String prepareBehaviour = jsonService.prepareBehaviour( behaviours.get( 0 ) );
            // TODO: map behaviours to JSON
            // List<JSONObject> collect = behaviours.stream().map(b -> new JSONObject(b)).collect(Collectors.toList());
            List<JsonNode> collect = behaviours.stream()
                    .map(b -> {
                        try {
                            MapperService.removeDBManagementFields((ObjectNode) b);
                            return b;
                        } catch (Exception e) {
                            log.error(e);
                            return null;
                        }
                    })
                    .collect(Collectors.toList());

            return ResponseEntity.ok(collect);
        }
    }

    public ResponseEntity<JsonNode> setBehaviours(
            Principal principal,
            String behavior) throws Exception {
        String userId = this.authUtils.getUsernameFromPrincipal(principal);
        String dbId = this.dbUtils.getCurrentCatalogForUser(userId);

        try (ODatabaseSession ignored = dbService.acquire(dbId)) {

            ObjectNode mappedBehavior = (ObjectNode) MapperService.getJsonNode(behavior);

            String rid = this.dbService.getRecordId("Behaviours", mappedBehavior.get("_id").asText());
            if (rid != null) {
                mappedBehavior.put("@rid", rid);
            }

            JsonNode doc = this.dbService.save("Behaviours", rid, mappedBehavior.toString());
            MapperService.removeDBManagementFields((ObjectNode) doc);

            return ResponseEntity.ok(doc);

        }
    }

}

File: server/src/main/java/de/ingrid/igeserver/api/BehavioursApiController.kt
/**
 * ==================================================
 * Copyright (C) 2023 wemove digital solutions GmbH
 * ==================================================
 * Licensed under the EUPL, Version 1.2 or – as soon they will be
 * approved by the European Commission - subsequent versions of the
 * EUPL (the "Licence");
 *
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the Licence is distributed on an "AS IS" basis,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and
 * limitations under the Licence.
 */
package de.ingrid.igeserver.api

import de.ingrid.igeserver.persistence.postgresql.jpa.model.ige.Behaviour
import de.ingrid.igeserver.services.BehaviourService
import de.ingrid.igeserver.services.CatalogService
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.security.Principal

@RestController
@RequestMapping(path = ["/api"])
class BehavioursApiController : BehavioursApi, Logging {

    @Autowired
    private lateinit var catalogService: CatalogService

    @Autowired
    private lateinit var behaviourService: BehaviourService

    override fun getBehaviours(principal: Principal): ResponseEntity<List<Behaviour>> {
        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)

        val result = this.behaviourService.get(catalogId)
        return ResponseEntity.ok(result)
    }

    override fun setBehaviours(
        principal: Principal,
        behaviours: List<Behaviour>
    ): ResponseEntity<Void> {

        val catalogId = catalogService.getCurrentCatalogForPrincipal(principal)

        behaviourService.save(catalogId, behaviours)
        return ResponseEntity.ok().build()
    }
}
